config/migrate/migrate.go

```diff
@@ -49,6 +61,92 @@ func ZeroToOne(cfg *config.Config) error {
 	}
 
 	cfg.Revision = 1
+
+	if err := safeWriteConfig(cfg); err != nil {
+		rollbackConfigWrite(cfg)
+		return err
+	}
+
+	return nil
+}
+
+// OneToTwo migrates a configuration from Revision 1 to Revision 2
+func OneToTwo(cfg *config.Config) error {
+
+	// TODO(ramfox): qfs migration
+
+	if err := oneToTwoConfig(cfg); err != nil {
+		return err
+	}
+	if err := safeWriteConfig(cfg); err != nil {
+		rollbackConfigWrite(cfg)
+		return err
+	}
+	
+	return nil
+}
+
+func oneToTwoConfig(cfg *config.Config) error {
+	if cfg.API != nil {
+		apiCfg := cfg.API
+		defaultApiCfg := config.DefaultAPI()
+		if apiCfg.Address == "" {
+			apiCfg.Address = defaultApiCfg.Address
+		}
+		if apiCfg.WebsocketAddress == "" {
+			apiCfg.WebsocketAddress = defaultApiCfg.WebsocketAddress
+		}
+	} else {
+		return qerr.New(fmt.Errorf("invalid config"), "config does not contain API configuration")
+	}
+
+	if cfg.RPC != nil {
+		defaultRpcCfg := config.DefaultRPC()
+		if cfg.RPC.Address == "" {
+			cfg.RPC.Address = defaultRpcCfg.Address
+		}
+	} else {
+		return qerr.New(fmt.Errorf("invalid config"), "config does not contain RPC configuration")
+	}
+
+	cfg.Revision = 2
+
+	if err := cfg.Validate(); err != nil {
+		return qerr.New(err, "config is invalid")
+	}
+
+	return nil
+}
+
+
+func rollbackConfigWrite(cfg *config.Config) {
+	cfgPath := cfg.Path()
+	if len(cfgPath) == 0 {
+		return
+	}
+	cfgPathComponents := strings.Split(cfgPath, "/")
```

Comment:
This won't work on Windows. Use `filepath.Dir` to get the path without the last component, and use `filepath.Join` to create the fullpath with the new basename. `filepath` is aware of Windows slashes, while `path` (and string splitting) is not.

Reply:
Good point.